### PR TITLE
Correct initial state for nested method nullability analysis

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/61516
Fixes https://github.com/dotnet/roslyn/issues/61964

Using the snippet from https://github.com/dotnet/roslyn/issues/61516 (included as first test in this PR), at some point during the analysis (which takes many steps due to lambda inference, method re-inference and lambda conversion) we would visit the property access (`plan. Schedule`). We'd check the slot for the property (see `if (this.State.HasValue(slot))` in  `VisitMemberAccess`) and we'd find that it had value (`NotNull`, ie. false-false in the BitVector).
That value was never set explicitly. Instead, it resulted from an incorrect initialization (too large) of the state/BitVector.

I'm not exactly sure why this problem resulted in the observed behavior (addition of unrelated code affecting nullability analysis).

As part of my investigation and understanding how we track slots for nested methods, I added some assertions that tighten the relationship between `Variables` and `LocalState`. Namely, their `Id` and sizes should match.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -11337,7 +11344,7 @@ private static BitVector CreateBitVector(int capacity, bool reachable)
                 return state;
             }
 
-            private int Capacity => _state.Capacity / 2;
+            public int Capacity => _state.Capacity / 2;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:11347]: Is this change necessary?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: No longer. Thanks

**Review by 333fred — APPROVED** [review]

**Comment by cston** [comment]
>             // (false, false) => NotNull, (false, true) => MaybeNull, (true, true) => MaybeDefault.

Should we change `(false, false)` to represent "unset" rather than `NotNull` (and update `HasValue(int id, int index)` accordingly)? Would that prevent similar failures in the future?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:11286 in 9a5116e. [](commit_id = 9a5116e7b033c2968e8720611bc8437e52a224af, deletion_comment = False)

**Review by cston — APPROVED** [review]

File: src/Tools/Source/RunTests/TestRunner.cs [hunk 2]
```diff
@@ -127,7 +127,7 @@ internal async Task<RunAllResult> RunAllOnHelixAsync(ImmutableArray<WorkItemInfo
                 executable: _options.DotnetFilePath,
                 arguments: "build helix-tmp.csproj",
                 captureOutput: true,
-                onOutputDataReceived: (e) => ConsoleUtil.WriteLine(e.Data),
+                onOutputDataReceived: (e) => { Debug.Assert(e.Data is not null); ConsoleUtil.WriteLine(e.Data); },
```
Comment (jcouv) [inline src/Tools/Source/RunTests/TestRunner.cs:130]: Getting `##[error]src\Tools\Source\RunTests\TestRunner.cs(130,68): error CS8604: (NETCORE_ENGINEERING_TELEMETRY=Build) Possible null reference argument for parameter 'message' in 'void ConsoleUtil.WriteLine(string message)'.` in CI correctness leg.

If you look at the implementation of `CreateProcess`, the `onOutputDataReceived` callback is only invoked when `e.Data` isn't `null`.

**Review by cston — APPROVED** [review]
